_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,215 @@
+which `A` calls, we will get classloading exceptions when our code attempts to invoke the missing classes/methods.
+
+These are what we call **Binary Incompatibility Errors** -- errors that happen when the compiled bytecode references a name that cannot be resolved during runtime.
+
+Before we look at how to avoid binary incompatibility errors, let us first
+establish some key terminologies we will be using for the rest of the guide.
+
+## What are Evictions, Source Compatibility and Binary Compatibility?
+
+### Evictions
+When a class is needed during execution, the JVM classloader loads the first matching class file from the classpath (any other matching class files are ignored).
+Because of this, having multiple versions of the same library in the classpath is generally undesirable:
+
+* Need to fetch and bundle multiple library versions when only one is actually used
+* Unexpected runtime behavior if the order of class files changes
+
+Therefore, when resolving JARs to use for compilation and packaging, most build tools will pick only one version of each library and **evict** the rest.
+
+### Source Compatibility
+Two library versions are **Source Compatible** if switching one for the other does not incur any compile errors or unintended behavioral changes at runtime.  
+For example, If we can upgrade `v1.0.0` of a dependency to `v1.1.0` and recompile our code without any compilation errors, `v1.1.0` is source compatible with `v1.0.0`.
+
+### Binary Compatibility
+Two library versions are **Binary Compatible** if the compiled bytecode of these versions can be interchanged without causing binary compatibility errors.  
+For example, if we can replace the class files of a library's `v1.0.0` with the class files of `v1.1.0` without any binary compatibility errors during runtime,
+`v1.1.0` is binary compatible with `v1.0.0`.
+
+**NOTE:** While breaking source compatibility normally results in binary compatibility breakages as well, they are actually orthogonal -- breaking one does not imply breaking the other.
+
+#### Forwards and Backwards Compatibility
+
+There are two "directions" when we describe compatibility of a library release:
+
+**Backwards Compatible** means that a newer library version can be used in an environment where an older version is expected. When talking about binary and source compatibility,
+this is the common and implied direction.
+
+**Forwards Compatible** means that an older library can be used in an environment where a newer version is expected.  
+Forward compatibility is generally not upheld for userland libraries. It is only important in situations where an older version of a library is commonly used at runtime against code that is compiled with newer versions. (for example, [Scala's standard library](http://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html))
```

Comment:
Why/where is the Scala standard library more often replaced by an earlier version than other libraries?

Answer: it isn't. It's just that the Scala std lib has very broad exposure, and they want to support that feature for users who do put themselves in situations where they do that. However, those users would replace *other libraries* with earlier versions just as much as the Scala std lib. There is nothing special about the Scala std lib here.

The only well-known example in the "open" community around Scala is sbt: sbt loads itself with its declared version of Scala, and a bunch of dependencies (such as Zinc). It will then classloads sbt plugins, which will inherit from that hard-coded version of the Scala std lib *and any other dependency of sbt*, no matter what the sbt plugin was compiled with. (At least, I'm pretty sure that's what happens in sbt; I haven't actually tested that hypothesis. Maybe @jvican knows better.)

Reply:
Yes, I can confirm this. I brought it up in our last discussion of forward binary compatibility https://contributors.scala-lang.org/t/question-about-changes-in-the-scalac-release-cycle/1093/6. You explained it correctly. In reality, what happens is that global plugins and local plugins are classloaded in independent classloaders whose parent is the sbt classloader, but this does not change the behaviour that we're describing here.